charms/kserve-controller/src/charm.py

```diff
@@ -462,6 +469,35 @@ def gen_certs(self, namespace, service_name):
             ]
         )
 
+    def _push_controller_certificates(self):
+        """Push certificates to the kserve-controller workload container."""
+        self.controller_container.push(
+            "/tmp/k8s-webhook-server/serving-certs/tls.crt",
+            Path("/run/cert.pem").read_text(),
+            make_dirs=True,
+        )
+        self.controller_container.push(
+            "/tmp/k8s-webhook-server/serving-certs/tls.key",
+            Path("/run/server.key").read_text(),
+            make_dirs=True,
+        )
+
+    def _restart_controller_service(self) -> None:
+        """Restart the kserve-controller service."""
+        # If the kserve-controller service is not running, do nothing
+        try:
+            self.controller_container.get_service(self._controller_container_name).is_running()
+        except ModelError:
+            log.info("Service not found, nothing to restart.")
+            return
```

Comment:
Is this an event we're ok with just returning on?  Are there times where this occurs and we are ok with it?  I don't know when this will occur, but it feels like something that's a real error.  

If config-changed fires and we eventually hit this `except` block, what is the state of the charm?  Is it running and has the expected config?  I don't think so, because we haven't restarted it.  So this feels like an error or blocked state maybe?

Reply:
Ok I'm coming around to this a bit.

One way we hit this is the sequence `install -> start -> config-changed -> pebble ready`, right?  So in that case config-changed tries to upload the files but cannot because the container is not up.  Then later pebble-ready handler fires and that uploads the files anyway, and all is well.  Is that what you're thinking?

Actually relying on [update_layer](https://github.com/canonical/charmed-kubeflow-chisme/blob/main/src/charmed_kubeflow_chisme/pebble/_update_layer.py) here might be bugged too.  If a plan hasn't changed (eg: layer hasn't changed) I don't think `update_layer` triggers a restart of the service.  So changing the files might not restart things anyway

Reply:
I think what jumps out at me here is interaction between config-changed and the pebble container.  With this change, config_changed needs implementation details about the container (eg: that it has some files, etc).  Its encapsulated in helpers, but even still.

what about instead of doing 
```python
    def _on_config_changed(self, event):
        self._on_install(event)

        ################THIS STUFF
        if not self.controller_container.can_connect():
            self.unit.status = MaintenanceStatus("Waiting for kserve-controller to be reachable")
            event.defer()
            return
        self._restart_controller_service()
        ################THIS STUFF
```

instead just call the pebble_ready handler:
```python
    def _on_config_changed(self, event):
        self._on_install(event)

        # Refresh the credential files
        # could also do a self.on.pebble_ready.emit()
        self._on_kserve_controller_ready(event)
```

does this feel more readable?  It at least keeps config-changed and pebble-ready a bit more separated.  

As is, this will still break though because pebble_ready doesn't refresh the service if the files changed and layer stays the same.  Still need a check for that

Reply:
> One way we hit this is the sequence install -> start -> config-changed -> pebble ready, right? So in that case config-changed tries to upload the files but cannot because the container is not up. Then later pebble-ready handler fires and that uploads the files anyway, and all is well. Is that what you're thinking?

Oh, I'm sorry, but this does not have to do with any files. This piece of code just restarts the service, the layer is left unchanged. The workload container doesn't have any files to update on config changed.

> instead just call the pebble_ready handler:

I'd do that if I cared about the layer and all the pebble ready event handler does, but I only care about the service itself. Plus, calling `self._on_kserve_controller_ready(event)` wont' guarantee that the service is restarted as I expect since the internals of that event handler won't allow restarts if there are no changes in the pebble layer.